### PR TITLE
Implement WAI ARIA recommended focus handling

### LINKED ISSUE #354: Is it possible to block Tab (tabindex)
Hello everyone, i am making a form which contains rect-calendar, but when i am navigating through the form with Tabulation like this:

image

it enters the calendar and navigate to all the controls inside the calendar, i need to skip the calendar tabs and jump to my other input.

Is it actually possible in react-calendar? or do you know any other solution for this?

### PR DESCRIPTION
**Pull request by pekala**
Fixes #354 

Introduces the concept of an active focusable element, to adjust the keyboard interaction to match [WAI-ARIA recommendations](https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/datepicker-dialog.html). 

Kapture 2020-10-26 at 21 36 55

@wojtekmaj Would love your help in pushing this out. Things I don't handle yet (that I know of):

- [ ] Min/max dates (should be easy)
- [ ] Disabled dates - that one will be tricky. I guess ideally we would jump to the next non-disabled date, but not sure how we can implement that in a sane way

### CODE REVIEW

File: src/Calendar/Navigation.jsx [hunk 1]
```diff
@@ -78,6 +78,14 @@ export default function Navigation({
     && maxDate <= nextActiveStartDate2
   );
 
+  // Make sure the navigation is not navigable at the first render
+  // so that the calendar takes the initial focus.
+  const [tabIndex, setTabIndex] = React.useState(-1);
+  React.useEffect(() => {
+    setTabIndex(-1);
+    setTimeout(() => setTabIndex(0), 0);
+  }, [view]);
```
Comment (wojtekmaj) [inline src/Calendar/Navigation.jsx:87]: That's a breaking change - we're currently requiring React 16.3 and up. Not saying this is a bad thing, just noting this to self.

Comment (pekala) [reply src/Calendar/Navigation.jsx:87]: Ah right. Well, you could make that call I guess, in the similar spirit as not supporting older browsers.

**Comment by pekala** [comment]
@wojtekmaj so, WDYT? Is this an acceptable path forward?

**Comment by wojtekmaj** [comment]
Yes, absolutely! We will need to wait for the next major release with this though, unfortunately.

**Comment by thepuzzlemaster** [comment]
You may have already addressed it, but I could see an argument made for still being able to focus a disabled element. [W3 guidance](https://www.w3.org/TR/wai-aria-practices-1.1/#Listbox) around disabled elements within a listbox, for instance support focusing on disabled elements, as otherwise, a non-sighted user would not be aware of them.

Granted, [grids](https://www.w3.org/TR/wai-aria-practices-1.1/#grid), are not clear on if a cell is disabled, how that would effect keyboard navigation/focus - I could see an argument made here.

**Comment by skatt931** [comment]
Hello, do you have an approximate time for when this feature will be released?

**Comment by wojtekmaj** [comment]
This change is breaking compatibility with older versions of React so this will need to be merged after the next major, dropping support for React <16.8 is released.

**Comment by clicktravel-danielkentfield** [comment]
Is there any info on when this might be released?

This is quite an important feature for us.

**Comment by Merri** [comment]
This focus handling implementation won't work when rendering multiple months:

Multiple months

We don't want the calendar component to automatically shift `activeStartDate` within a single calendar instance, but would like to see it jump to the next calendar.

Doing this multiple months feature is pretty easy with current `react-calendar`:

```tsx
export function MonthlyCalendar(props: Props) {
  const { formatDate } = useIntl();
  // date-fns
  const months = eachMonthOfInterval({ start: props.minDate, end: props.maxDate });

  return (
    <>
      {months.map((startOfMonth) => (
        <Wrapper key={startOfMonth.toISOString()}>
          <Title>
            <FormattedDate value={startOfMonth} month="long" year="numeric" />
          </Title>
          <Calendar
            {...props}
            activeStartDate={startOfMonth}
            maxDetail="month"
            showNavigation={false}
            showNeighboringMonth={false}
            formatLongDate={(_locale, date) => formatDate(date, dayOptions)}
          />
        </Wrapper>
      ))}
    </>
  );
}
```

It might be that this kind of rendering has to be supported by `react-calendar` to make the focus handling work as expected in this use case?

**Comment by deanohyeah** [comment]
can we bring this back now that 4.0.0 is out with react 16.8+

**Comment by wojtekmaj** [comment]
We most definitely can!

**Comment by wojtekmaj** [comment]
Some issues I have noticed:
- Using arrows when focus is on calendar tiles causes the page to scroll
- I can't seem to be able to navigate to the tiles - I feel like after you tabbed through navigation the focus should go to the tile grid?

**Comment by pekala** [comment]
Unfortunately I won't find time to work on this any time soon :(

**Comment by thepuzzlemaster** [comment]
I'd love to help get this over the finish line, as it seems pretty close. Thanks for the great start @pekala.
I've pulled down the most recent working branch from your repo, and rebased on top of the latest from main in this repo and pushed it up [here](https://github.com/thepuzzlemaster/react-calendar/tree/aria-navigation).

I fixed the issue with the page scrolling on arrow usage.
I see a couple other things I'd like to improve here, but I wanted to check on a couple things first.

1. I noticed that there are a bunch of test failures when running `yarn test`. Have the tests ever passed? I'm not sure if this was a change related to rebasing on top of the updates in `main`, or if they just never got run, and I need to make some fixes to address the changes.
2. Are @pekala and @wojtekmaj okay with me taking the baton in getting this over the finish line? Seems like everyone would be on board, but just wanted to make sure.
3. Should I just create a new PR, or is there a way to edit this PR to point to my fork now instead, so we can continue the conversation here?

Thanks!

**Comment by thepuzzlemaster** [comment]
I ended up creating a new PR - figuring they will reference each other, so none of the history in this thread will be lost.

**Pull request closed without merging**